### ISSUE TITLE
Clarify relation and difference to parent OneOf repo

### ISSUE THREAD

**Issue by miniyou** [DiscU/DiscU#4]
Stumbled upon your repo when researching discriminated unions for C#. Great effort!

Since both this repo and the parent (https://github.com/mcintyre321/OneOf) seem to be actively maintained, but also diverging (`This branch is 174 commits ahead, 78 commits behind mcintyre321:master`), could you clarify in the readme what the goal of this fork is?

After a cursory look, the main difference seems to be `Else` and `ElseThrow`, but it's hard to tell.

Are you planning to upstream your contributions? Do you have other goals?

**Comment by ghost** [comment]
Thanks for your feedback Hampus.

The initial fork of OneOf was an experiment of mine to see if the code
generation could be made easier, plus tweaks to the API for ease of use.  A
little down the track, James jumped in and largely took over, making major
changes to the generated code which is now designed to result in
compile-time errors instead of runtime errors if you forget to include an
option in your switch or match, unless you use Else (etc) -- it was a
pretty nifty design change.  We contacted Harry (OneOf author) some time
back to see if he was interested in a PR, but he declined the changes, so
DiscU was born to differentiate the two going forward.

I always hoped that C# would evolve to the point that DiscU won't be needed
(but James might have other ideas!).  C# 7's improved switch statement and
C# 8's new switch expressions looks like they might have finally brought us
to this point.   I consider that a win.

James, what are your thoughts with regards to DiscU now that C# has matured
enough to support it natively?  (for better or worse)

**Comment by jamesbascle** [comment]
The big difference between DiscU and the type cases added in c# 7 is that
this lets you create strict sumtypes and only allows other types to be
referenced by the OneOf object in if they are added to the generic
parameters of the one of.  There are advantages and disadvantages of
course. but I think we did a pretty good job.

The difference between this and OneOf is that we use a fluent builder
syntax rather than just having the user pass a bunch of functions in.  This
allows for the user to write the code in the order they would like, rather
than the order in which they have written the generic arguments.
We also added support for passing in a reference to a child class as the
parent class (if the parent class is in the Union) and it will call the
proper subclass case rather than the parent class case.  Overall it's a
more flexible syntax than the one in OneOf.  There is a small performance
hit, but I believe we have sped it up to nearly the limit without writing
bare IL.  It's very fast according to our performance tests.

A while ago, I was fooling around with the DSL to see if I could add more
strictness to certain constructs, or extend it at all, but due to the
factorial nature of the algorithm that generates the code, and the
difficulty of extending some of the other constructs.  I think it will stay
as is for now.  Someday I may get the inspiration for some way to improve
it, but I think it's feature complete at this point.
